o16n: primitives code location

alxhub opened the pull request:
Nothing to see here, Enea

pkozlowski-opensource committed:
fix(core): drop mutate function from the signals public API (#51821)

This change removes the `mutate` method from the `WritableSignal` interface and
completely drops it from the public API surface.

The initial API proposal for Angular signals included the mutate method, allowing
in-place modification of JS objects, without changing their references (identity).
This was based on the reasoning that identity change on modification is not necessary
as we can send the “modified” notification through the signals graph.
Unfortunately the signal-specific change notification is lost as soon as we read
signal value outside of a reactive context (outside of a reactive graph).
In other words - any code outside of the Angular signals library can’t know
that an object is modified.

Secondly, to make the mutate method work, we’ve defaulted the signal value equality function
to the one that considers non-primitive values as always different.
This is unfortunate for people working with immutable data structures
(this is notably the case for the popular state management libraries)
as the default equality function de-optimizes memoization in computed,
making the application less performant.

Given the above reasons we prefer to remove the mutate method in the signals library -
at least for now. There are just too many sharp edges and tradeoffs that we don’t fully
understand yet.

BREAKING CHANGE:

The  `mutate` method was removed from the `WritableSignal` interface and completely
dropped from the public API surface. As an alternative please use the update method and
make immutable changes to the object.

Example before:

```typescript
items.mutate(itemsArray => itemsArray.push(newItem));
```

Example after:

```typescript
items.update(itemsArray => [itemsArray, …newItem]);
```

alxhub committed:
refactor(core): extract signals API away from the 'signals' package

This commit refactors the signals API surface of Angular out of the
//packages/core/src/signals package. This is done in preparation of moving
the core signals package into a new 'primitives' package that's decoupled
from the public API.

alxhub committed:
refactor: move signals code into primitives package

This commit reorganizes the Angular code a bit, and moves signals into a
newly defined `@angular/core/primitives` location. This will be used inside
g3 to allow non-Angular targets to depend on the signals core without
incurring a dependency on the whole framework.

eneajaho commented on packages/core/src/render3/reactivity/signal.ts:
```diff
@@ -0,0 +1,84 @@
+   * Update the current value by mutating it in-place, and
+   * notify any dependents.
+   */
+  mutate(mutatorFn: (value: T) => void): void;
```
So we are keeping the mutate fn? https://github.com/angular/angular/pull/51821

pkozlowski-opensource replied:
Probably not :-)

rhutchison commented on packages/core/src/render3/reactivity/signal.ts:
```diff
@@ -0,0 +1,83 @@
+  set(value: T): void;
+
+  /**
+   * Update the value of the signal based on itfs current value, and
```
itfs?

alxhub committed:
refactor(core): add `consumerOnSignalRead` hook to `ReactiveNode`

This hook allows a consumer to create `ReactiveNode`s with custom behavior
when signals are read.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

alxhub commented:
Caretaker: This will need cl/571389475 in the sync.

alxhub commented:
Caretaker:

* pullapprove disabled (code was already approved, partially in another PR)
* this change requires a g3 patch: cl/571389475
* the CI failure is due to the symbol test for defer which is already fixed

atscott commented:
This PR was merged into the repository by commit 00128e38538f12fe9bc72ede9b55149e0be5ead0.
